contrib/hadoop-store-builder/src/java/voldemort/store/readonly/fetcher/HdfsFetcher.java

```diff
@@ -245,21 +255,67 @@ private File fetchFromSource(String sourceFileUrl,
                                                                  stats,
                                                                  status,
                                                                  bufferSize);
-            if(!fs.isFile(path)) {
+            if(!fs.isFile(rootPath)) { // We are asked to fetch a directory
                 Utils.mkdirs(destination);
-                HdfsDirectory directory = new HdfsDirectory(fs, path);
+                HdfsDirectory rootDirectory = new HdfsDirectory(fs, rootPath);
+                List<HdfsDirectory> directoriesToFetch = Lists.newArrayList();
 
-                HdfsFile metadataFile = directory.getMetadataFile();
+                HdfsFile metadataFile = rootDirectory.getMetadataFile();
                 Long estimatedDiskSize = -1L;
 
                 if(metadataFile != null) {
                     File copyLocation = new File(destination, metadataFile.getPath().getName());
                     fetchStrategy.fetch(metadataFile, copyLocation, null);
-                    directory.initializeMetadata(copyLocation);
-                    String diskSizeInBytes = (String) directory.getMetadata()
-                                                               .get(ReadOnlyStorageMetadata.DISK_SIZE_IN_BYTES);
-                    estimatedDiskSize = (diskSizeInBytes != null && diskSizeInBytes != "") ? (Long.parseLong(diskSizeInBytes))
-                                                                                          : -1L;
+                    rootDirectory.initializeMetadata(copyLocation);
+
+                    if (metadataFile.getDiskFileName().equals(ReadOnlyUtils.FULL_STORE_METADATA_FILE)) {
+                        // Then we are in build.primary.replica.only mode, and we need to determine which
+                        // partition sub-directories to download
+                        StoreDefinition storeDefinition = metadataStore.getStoreDef(storeName);
+                        Cluster cluster = metadataStore.getCluster();
+                        List<Integer> partitionsMasteredByCurrentNode = cluster.getNodeById(metadataStore.getNodeId()).getPartitionIds();
+                        RoutingStrategy routingStrategy = new RoutingStrategyFactory().updateRoutingStrategy(storeDefinition, cluster);
+                        for (int partitionId = 0; partitionId < cluster.getNumberOfPartitions(); partitionId++) {
+                            // For each partition in the cluster, determine if it needs to be served by the current node
+                            boolean partitionIdIsServedByCurrentNode = false;
```

Comment:
I agree. I think it belongs in the ```RoutingStrategy``` implementation, but I'm not exactly sure where would be the best place to put it.